web/packages/teleport/src/Roles/RoleEditor/YamlEditor.tsx

```diff
@@ -39,13 +42,33 @@ export const YamlEditor = ({
   yamlEditorModel,
   onChange,
   onSave,
+  onPreview,
   onCancel,
 }: YamlEditorProps) => {
   const isEditing = !!originalRole;
+  const [hasPreviewed, setHasPreviewed] = useState(!onPreview);
 
   const handleSave = () => onSave?.(yamlEditorModel.content);
 
+  // this returns a function or undefined, determined by the existence
+  // of the onPreview callback. onPreview is only available on clusters
+  // with policyEnabled.
+  // The only thing we need to do in this method besides calling the passed in onPreview
+  // is set state to know if the user has previewed the current change or not.
+  const getHandlePreview = () => {
+    if (onPreview) {
+      return () => {
+        onPreview();
+        setHasPreviewed(true);
+      };
+    }
+    return undefined;
+  };
```

Comment:
Instead of this being a function that you call to either get a function or undefined back, can't this just be a function that is passed down to `onPreview`, which does nothing if `onPreview` is not true?

Reply:
You could pass down `onPreview`, or make the prop `onPreview={onPreview ? getHandlePreview : null}`

Either approaches make it just a bit more obvious what's going on

Reply:
> onPreview={onPreview ? getHandlePreview : null}

whats funny is i had this in my implementation before the PR and thought "oh man, thats hard to read" haha. will go back to that, i think its better tbh